source/agora/script/Opcodes.d

```diff
@@ -0,0 +1,167 @@
+/*******************************************************************************
+
+    Contains the supported opcodes for the basic execution engine (non-webASM)
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.script.Opcodes;
+
+import std.traits : EnumMembers;
+
+/// The supported opcodes
+/// Opcodes named `CHECK_*` push their result to the stack,
+/// whereas `VERIFY_*` opcodes invalidate the transaction if the result is false.
+/// Can encode up to 255 opcodes (one of which is INVALID).
+/// Note that the range between `PUSH_BYTES_1` and `PUSH_BYTES_64` is
+/// purposefully reserved to encode a push to the stack between 1 .. 64 bytes,
+/// without having to encode it in a separate length byte. 64 was chosen as
+/// the upper bound because our signatures are 64 bytes.
+/// For pushes of data longer than 64 bytes use the `PUSH_DATA_*` opcodes.
+enum OP : ubyte
+{
+    /// Using this is an error and will invalidate the transaction
+    /// Purposefully located first to default OPs to errors.
+    INVALID = 0x45,
+
+    /// Used to encode small length of data to push to the stack (up to 64 bytes),
+    /// may be used with `case PUSH_BYTES_1: .. case PUSH_BYTES_64:` syntax.
+    PUSH_BYTES_1 = 0x01,
+    PUSH_BYTES_64 = 0x40, // 64 decimal
+
+    /// The next 1 byte contains the number of bytes to push onto the stack
+    PUSH_DATA_1 = 0x41,
+
+    /// The next 2 bytes (ushort in LE format) contains the number of bytes to
+    /// push onto the stack
+    PUSH_DATA_2 = 0x42,
+
+    /// The next 4 bytes (ushort in LE format) contains the number of bytes to
+    /// push onto the stack
+    RESERVED_PUSH_DATA_3 = 0x43,
```

Comment:
One question remains. How do we maintain backward compatibility when introducing new opcodes?

Bitcoin does it by causing the reserved opcodes to become a simple push to the stack. And then when the opcode is later defined to do something in a soft-fork, the redeem script actually has to use a `OP_DROP` opcode first.

There's more info here: https://bitcoin.stackexchange.com/a/85472/93682

I will likely work on this another time, I don't want to stall the PR.